### PR TITLE
[HZ-380] Cache merkle tree support

### PR DESCRIPTION
**Pull request by ahmetmircik**
ee counterpart: https://github.com/hazelcast/hazelcast-enterprise/pull/4097

__Modifications:__
Introduced `CacheRSMutationObserver` to observe mutations in cache record stores. Implemented it for event-journal and merkle-tree support.

*Also includes client protocol changes*

### CODE REVIEW

**Comment by ahmetmircik** [comment]
run-lab-run

**Review by blazember — COMMENTED** [review]
The declarative config-related changes are missing, but otherwise LGTM.

File: hazelcast/src/main/java/com/hazelcast/config/CacheConfig.java [hunk 1]
```diff
@@ -519,6 +543,10 @@ public void writeData(ObjectDataOutput out) throws IOException {
         out.writeBoolean(disablePerEntryInvalidationEvents);
 
         writePartitionLostListenerConfigs(out);
+
+        if (out.getVersion().isGreaterOrEqual(Versions.V5_0)) {
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/config/CacheConfig.java:547]: nit: mark this and other rolling-upgrade related changes with a `// RU_COMPAT_4_2` comment so they can be easily spotted

**Review by vbekiaris — COMMENTED** [review]
Looks good. We are missing a change in `CacheConfigHolder` to support merkle-tree configuration from client side. Also, since JCache API supports dynamic creation of caches, it would be great to verify caches created with `CacheManager#createCache` from member or client-side provider have a properly applied `MerkleTreeConfig`.

File: hazelcast/src/main/java/com/hazelcast/config/CacheConfig.java (hunk 1 above)
Comment (ahmetmircik) [reply hazelcast/src/main/java/com/hazelcast/config/CacheConfig.java:548]: done

**Review by blazember — APPROVED** [review]
We should verify the generator changes in `ConfigXmlGeneratorTest`, but otherwise LGTM :+1:

**Review by vbekiaris — APPROVED** [review]

**Review by sancar — APPROVED** [review]

**Pull request merged**